#5 feat: 협찬하기 UI 구현

loading1031 opened the pull request:
## #️⃣ Issue Number

#5 

## 📝 요약(Summary)

<!--- 변경 사항 및 관련 이슈에 대해 간단하게 작성해주세요. 어떻게보다 무엇을 왜 수정했는지 설명해주세요. -->
1. 협찬하기 ui 구현: 상태 관리
2. theme.color 색상 추가: hint, main-blue 등
3. vite 보안취약점에 따른 버전 업데이트: 6.2.0 -> 6.2.6
4. 피그마에서 텍스트로 표시하던 것들을 협찬 상태(작성중, 작성 완료)에 따라 `자물쇠`,`휴지통`, ` 체크`등의 아이콘으로 변경하였습니다.
5. 프롬프트 입력시, 줄바꿈이 되도록 설정하였습니다.
6. 헤더 높이를 전역 상태관리하였습니다.


## 🛠️ PR 유형

어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [x] 버그 수정
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [x] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## 📸스크린샷 (선택)
<img width="1437" alt="image" src="https://github.com/user-attachments/assets/e6c96955-bac0-4876-bb69-0219f362714d" />

## 💬 공유사항 to 리뷰어

<!--- 리뷰어가 중점적으로 봐줬으면 좋겠는 부분이 있으면 적어주세요. -->
- Figma에서 사용되는 색상 중 일부는 `theme` 폴더에 등록했습니다!
- FSD에 맞게 잘 설계한건진 잘 모르겠군요,, 

<!--- 논의해야할 부분이 있다면 적어주세요.-->
- 메뉴탭 클릭시 이동은 아직 안했습니다... 로그인 후에 role 값으로 이동하도록 수정하겠습니다
- 협찬 제출폼 디자인이 미흡합니다. 나중에 디자인만 따로 수정해봐야겠습니다
<!--- ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요? -->

## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

odukong committed:
feat : 공통 Title 컴포넌트 구현

- 모든 페이지에서 공통적으로 사용되는 Title 컴포넌트를 구현하였습니다.

ref : #6

odukong committed:
feat : MY카드 페이지 UI 구현

-  비밀번호 일치여부에 따라 UserMyCard 페이지에서 PasswordAuth, MyCardInfo 컴포넌트를 구분하여 렌더링됩니다.

ref : #6

odukong committed:
style : 메뉴바 이동 링크 수정

- MY카드 메뉴의 이동링크를 /card에서  /mycard로 수정하였습니다.

ref : #6

odukong committed:
feat : 마이페이지 UI 구현

- 마이페이지 정보 보기 UI
- 마이페이지 정보 수정 UI

ref : #6

odukong committed:
design :  Modal padding 값 수정

- 모달 컴포넌트의 padding값을 동적으로 변경될 수 있도록 수정하였습니다.

ref : #6

loading1031 committed:
feat: 헤더 전역 상태 관리 및 theme 색상 추가

loading1031 committed:
feat: 프롬프트 컴포넌트 구현 및 협찬하기 페이지 라우팅

odukong committed:
feat : 공통 Table 컴포넌트 구현

ref : #6

odukong committed:
feat : 공통 페이징 컴포넌트 구현

ref : #6

odukong committed:
feat : 유저 결제내역 페이지 구현

ref : #6

loading1031 committed:
feat: 입력 및 메시지 줄바꿈 기능 추가
- AI 아이콘 추가

loading1031 committed:
feat: 혜택 작성 버튼 추가

odukong committed:
feat : 혜택 히스토리 UI 구현

- 사용가능한 혜택 리스트인 AvailableList 컴포넌트를 구현하였습니다.
- 혜택 모아보기 페이지인 UserBenefit UI를 구현하였습니다.

ref : #6

loading1031 committed:
feat: 협찬작성 컴포넌트 UI 생성
- 한글 마지막 글자 중복 입력되는 버그 수정

loading1031 committed:
feat: 협찬 내역 좌측 사이드 바 ui 구현
- 클릭한 혜택내역 상태관리
- main_blue, main_gray 추가

loading1031 committed:
refactor: 혜택 상태 관리 리스트로 변경
- DataForm -> ResponseDTO
- 채팅방 이동시, Popover 닫히도록 변경
ref : #4

loading1031 committed:
refactor: 혜택 상태값에 따라 버튼 변경
- COMPLETED, 모든 버튼 비활성화
- COMPLETED, 제출완료 버튼 추가
- COMPLETED, 제출/저장 버튼 제거
- PENDING, 삭제 아이콘 버튼 추가

ref : #4

loading1031 committed:
fix: vite 보안취약점 수정
- 깃허브 시큐리티 취약점 알림
- Vite has an `server.fs.deny` bypass with an invalid `request-target` #1

loading1031 committed:
fix: activeBenefit 초기값 설정
- sponsor 화면에 처음 진입시 혜택 작성버튼 클릭 안되는 버그 수정

loading1031 committed:
Merge branch 'dev' into feat/#4/협찬하기-UI

loading1031 commented on src/features/sponsor/type/converter.ts:
```diff
@@ -0,0 +1,16 @@
+  return formData;
+}
+
+export const benefitFormToRequestDTO = (
```
컨버터 함수명이 고민이 되네요

odukong replied:
BenefitFormData 타입으로 형식을 바꾸는 거라면 `convertToBenefitFormData` 정도로만 바꿔도 좋을것 같아요! 
개인 의견이니 참고만 해주세요!

loading1031 replied:
함수니까 동사를 쓰는게 맞는거 같습니다~~ 수정하겠습니다

odukong commented on src/pages/Sponsor.tsx:
```diff
@@ -0,0 +1,140 @@
+  text: string;
+}
+
+const Sponsor = () => {
```
협찬사 기능이 **프롬프트** 제외하고도, 발행내역 페이지도 존재하니 `SponsorPrompt` 정도로 컴포넌트명을 수정하면 좋을 것 같습니다.

loading1031 replied:
`협찬하기` 페이지가 영어로 `sponsor`더라구요. 
페이지명에 따라 한거긴 한데, 명사랑도 겹치니.. `SponsorPrompt`로 정정하겠습니다

odukong commented on src/pages/Sponsor.tsx:
```diff
@@ -0,0 +1,140 @@
+      if (activeBenefitId === null) return;
+
+      setBenefitList((prev) =>
+        prev.map((benefit) =>
+          benefit.benefitId === activeBenefitId ? { ...benefit, [field]: value } : benefit,
+        ),
```
map 메서드의 경우, 조건에 일치하는 값을 찾았다 하더라도 전체 배열을 순회하게 되면서 불필요한 순회가 이루어질 수 있는 단점이 있습니당. 현재 로직 상으로 `benefitList`에서 `activeBenefitId`와 일치하는 하나의 요소만 찾으면 되는 부분이기 때문에 아래와 같이  `setBenefitList` 변경 함수 내에서 `findIndex` 메서드를 통해 id가 일치하는 요소의 인덱스를 찾는 식으로 수정하면 불필요한 순회를 막을 수 있을 것 같아요!
```ts
const index = prev.findIndex((b) => b.benefitId === activeBenefitId); // id가 일치하는 배열의 인덱스 찾기
if (index === -1) return prev;

const updated = [...prev];
updated[index] = { ...updated[index], [field]: value };   // 배열의 특정 인덱스 부분만 값을 수정 
return updated;
```

loading1031 replied:
오 그러네요 수정하겠습니다~

odukong commented on src/pages/Sponsor.tsx:
```diff
@@ -0,0 +1,140 @@
+      );
+    };
+
+  const handleDateChange = (field: keyof BenefitFormData) => (date: Date | null) => {
+    if (!date || activeBenefitId === null) return;
+
+    setBenefitList((prev) =>
+      prev.map((benefit) =>
+        benefit.benefitId === activeBenefitId ? { ...benefit, [field]: date } : benefit,
+      ),
+    );
+  };
```
`handleDateChange` 함수와  `handleBenefitDataChange` 함수가 받아오는 인자 값에만 차이가 있고 동일한 로직을 가지고 있기 때문에 하나의 공통 핸들러로 분리하여 각 코드에 적용하면 더 깔끔한 코드가 될 것 같습니당.

loading1031 replied:
옙! 수정하겠습니다~

odukong commented on src/shared/store/index.ts:
```diff
@@ -1 +1,12 @@
 // index.ts
+import { create } from 'zustand';
+
+interface HeaderState {
+  headerHeight: number;
+  setHeaderHeight: (height: number) => void;
+}
+
+export const useHeaderStore = create<HeaderState>((set) => ({
+  headerHeight: 0,
+  setHeaderHeight: (height) => set({ headerHeight: height }),
+}));
```
**store/index.ts** 경로에는 만들어진 **store**를 모아두는 모음집?역할로 사용하는게 좋을 것 같아서
해당 store 파일은 별도 파일(ex. headerStore.ts) 로 분리한 후에, 아래와 같이 index.ts 파일에서 export 해주면 좋을 것 같아용 
```ts
// index.ts
export { useHeaderStore } from './headerStore'; 
```

loading1031 replied:
아하 export 용도군요 옙!

odukong reviewed (commented):
구조도 FSD 구조에 잘 맞춰서 구현하신 것 같습니다~!! 고생하셨습니당 ( •̀ ω •́ )✧

loading1031 committed:
refactor: store/index.ts 파일 분리
- index.ts는 export만 선언
- headerStore.ts에 기능 이전

ref: #5

loading1031 committed:
refactor: PR 리뷰 반영
- 컨버터 함수명 변경
- SponsorPrompt 페이지/컴포넌트 명 변경
- activeBenefit을 find로 찾도록 변경
- handleDate/Data 통합
- feature 폴더 구조 변경
- Benefit status 타입 변경

loading1031 committed:
refactor: 협찬하기 라우팅 변경
ref: #5
